### PR TITLE
#160934074 Implement auth/signup endpoint with tests

### PR DESCRIPTION
**Pull request by Bobsar0**
#### What does this PR do?
Implements auth/signup endpoint using TDD approach

#### Description of Task to be completed?
- Create validator class
- Test and implement email validator
- Test and implement password validator
- Test and implement hash password method
- Test and implement compare password method
- Test and implement generate token method
- Test and create user model class
- Test and create user controller class
- Test and implement auth/signup endpoint on user controller class
- Add server route to endpoint

#### How should this be manually tested?
  Clone the  repo and run the following commands:
``` 
  git checkout ft-api-auth/signup-#160856655
  npm test
  npm start
 ```
Test api endpoint: POST/  http://localhost:5000/api/v1/auth/signup/ with curl or POSTMAN 

#### What are the relevant pivotal tracker stories?
https://www.pivotaltracker.com/story/show/#160856655

### CODE REVIEW

File: server/api/controllers/usersController.js [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+    this.user = userModel;
+  }
+
+  /**
+     * Create A User
+     * @param {object} req
+     * @returns {object} user object
+     */
+  async create(req) {
+    if (!req.body.username || !req.body.email || !req.body.password) {
```
Comment (robocopkaka) [inline server/api/controllers/usersController.js]: Can you use the object destructuring to get the values in `req.body` and use the destructured values throughout the function

File: server/api/models/userValidator.js [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    } if (this.password.length < 6) {
+      return 'Your password must be composed of at least 6 characters';
+    }
+    return 'true';
+  }
+
+  /**
+   * Hash Password Method
+   * @returns {string} hashed password
+   */
```
Comment (robocopkaka) [inline server/api/models/userValidator.js]: Please move functions used for authentication/authorization to a different file. Functions in this file should only validate inputs

**Comment by robocopkaka** [comment]
Also, could you test this endpoint - `${prefix}/auth/signup`. Know you've written tests for the methods that make it tick. Testing the endpoint directly ensures that all the methods that the endpoint interacts with are hit and removes the need to test each method individually.

File: server/api/controllers/usersController.js (hunk 1 above)
Comment (Bobsar0) [reply server/api/controllers/usersController.js]: Ok done. Please review

File: server/api/models/userValidator.js (hunk 2 above)
Comment (Bobsar0) [reply server/api/models/userValidator.js]: Okay done. Please review

**Pull request merged**